File: parser.rs
use std::fs;
#[path = "encoder.rs"]
mod encoder;

pub struct Storage {
    // The lookup table:
    // A table which stores vectors of three i32s
    // in the following format:
    // [
    //      Numerical representation of a datatype (0-4),
    //      Start index in raw table of data,
    //      End index in raw table of data
    // ]
    pub lookup: Vec<Vec<i32>>,

    // Stores any referenced value in the form
    // of binary as unsigned 8-bit integers.
    // When a value is stored, it will be converted
    // to binary and added here byte by byte. The
    // start and end index of this chunk of data
    // will then be appended to the lookup table.
    // This data will be decoded if it needs to
    // be used.
    pub raw: Vec<u8>
}



fn inform(message: String) {
    eprintln!("\x1b[31m[Bronze] \x1b[92m{}\x1b[0m", message.as_str());
}

fn raise(line: u8, message: String) {
    eprintln!("\x1b[31m\n[Bronze] \x1b[91mError at line {}: {}\n\x1b[0m", line, message.as_str());
    std::process::exit(1);
}

fn warn(line: u8, message: String) {
    eprintln!("\x1b[31m\n[Bronze] \x1b[93mWarning at line {}: {}\n\x1b[0m", line, message.as_str());
}


// Takes a script file and parses it into an array of binary which will be easier to process later
pub fn parse(file: &str) -> (Vec<Vec<Vec<u8>>>, Storage) {
    let file = match fs::read_to_string(file) {
        Ok(r) => {
            inform(
                format!("Read script file \"{}\" successfully", file)
            ); r
        },
        
        Err(_e) => {
            raise(
                0 as u8, 
                format!("Could not read \"{}\"", file)
            ); 
            String::from("No content")
        }
    };

    // Explicit storage
    let mut expl_storage = Storage {
        lookup: Vec::new(),
        raw: Vec::new()
    };

    let mut parsed: Vec<Vec<Vec<u8>>> = Vec::new();
    let mut line: Vec<Vec<u8>> = Vec::new();
    let mut expl_binary: Vec<u8> = Vec::new();
    let mut chunk = String::new();
    let mut chunk_type = 0;

    let mut ignore_delimiters: bool = false;
    let mut comment_ignore: bool = false;

    let mut line_started: bool = true;

    let mut line_count: u8 = 1;
    //line.push(1);

    // These are unrelated variables which are
    // referenced and changed throughout the loop
    let mut raw_start = 0;

    let mut parsed_integer = 0;
    let mut leading_zero: bool = false;
    // ==========================================

    for c in file.chars() {
        // Ignore newlines
        if c == 0x0A as char {
            line_count += 1;
            continue;
        }

        // Toggle comment_ignore if ` is present
        if c == 0x0060 as char {
            comment_ignore = !comment_ignore;
            continue;
        }
        if comment_ignore {
            continue;
        }

        // Enable/disable ignore_delimiters if parentheses are present
        if c == 0x028 as char {
            ignore_delimiters = true;
            continue;
        } else if c == 0x029 as char {
            ignore_delimiters = false;
            continue;
        }
        
        // Ignore whitespaces at the beginning of a line
        if c.is_whitespace() {
            if line_started {
                continue;
            }
        } else {
            line_started = false;
        }


        if !ignore_delimiters {
            // Colons separate a type declaration from
            // the actual value. The syntax of a given
            // value to a command should be written as
            // the type followed by a colon and then the
            // value (e.g. int:4). If there is no type
            // declaration, the value is interpreted
            // as a string by default.
            if c == 0x003A as char {
                // Since the "chunk" variable is built up
                // character by character, if a colon is found
                // the chunk variable must contain the type name.

                chunk_type = match chunk.as_str() {
                    "str" => 0, // String (0) (declaration optional)
                    "int" => 1, // Integer (1)
                    "float" => 2, // Float (2)
                    "bool" => 3, // Boolean (3)
                    _ => 4, // Default to string (0)
                };

                // Since the chunk is built char by char, Clearing the chunk
                // at this point will make it so that when the delimiter code
                // below is triggered, the chunk will not contain the declaration
                // but the raw value only.
                chunk = String::new();

                continue;
            }

            // Delimiters are spaces and semicolons
            // A semicolon represents the end of a line
            // A space represents the end of a chunk unless between parentheses
            // This checks if the char is one of these delimiters: 
            if
                c == 0x003B as char || // semicolon
                c == 0x20 as char // space
            {
                // If a delimiter has been found, it means
                // the current chunk of the line should be processed
                // and added to the line vector.

                println!("{chunk_type} {chunk}");

                // Add value to storage
                raw_start = expl_binary.len() as i32;
                expl_binary.push(chunk_type);
                match chunk_type {
                    0 => {encoder::enc_ascii(&mut expl_binary, chunk)}, // String (ASCII)

                    // Type is an integer
                    1 => {
                        // Parse to largest possible size
                        parsed_integer = match chunk.parse::<i128>() {
                            Ok(i) => i,
                            Err(e) => {
                                raise(
                                    line_count, format!(
                                        "The given value \"{}\" cannot be interpreted as a 128-bit integer.", chunk
                                    )
                                ); 0
                            }
                        };

                        encoder::enc_int(&mut expl_binary, parsed_integer)
                    },

                    _ => {}
                }

                line.push(expl_binary);
                expl_binary = Vec::new();

                // Add indices to lookup table
                //expl_storage.lookup.push(
                //    vec![
                //        chunk_type, // Numerical representation of datatype
                //        raw_start, // Start index in raw storage
                //        expl_binary.len() as i32 - 1 // End index in raw storage
                //    ]
                //);
                
                // Assign new string to hold next chunk
                chunk = String::new();
                chunk_type = 0;
    
                // Append line to parsed if
                // the delimiter is a semicolon
                if c == 0x003B as char {
                    parsed.push(line);
                    line = Vec::new();
                    // Add line number for debugging purposes
                    //line.push(line_count);

                    line_started = true;
                }
    
                continue;
            }
        }

        chunk.push(c);
    }

    return (parsed, expl_storage);
}

File: parser.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#include "encoder.c"

size_t strlen(const char *str)
{
    const char *s;
    for (s = str; *s; ++s);
    return(s - str);
}

// Read and refine script files
char*** parse(char* filename) {
    // Read script file
    FILE* fptr;
    fptr = fopen(filename, "r");

    FILE* out;
    out = fopen("out.c", "a");


    char* lineSentinel = malloc(5);
    lineSentinel = "LEND";


    // Will store pointers for each line
    char*** lines = NULL; // Where's my award?
    int lineCount = 0;

    // This will be toggled if the reader comes
    // across a comment symbol (`)
    bool comment_ignore = false;

    // This will be toggled if the reader comes
    // across parentheses, telling it to ignore spaces in-between them.
    bool ignore_delimiters = false;
    
    bool line_started = true;
    int chunk_encoding = 0; // Defaults to utf-8 (0)
    
    int segment_index = 0; // Index of current segment of current line being written to
    int segment_mem = 8; // Default memory given to a segment
    int segment_len = 0;
    char* segment_buffer = malloc(segment_mem);

    char** lineBuffer = NULL; //malloc(lineMemory);
    char ch;
    while (ch != EOF) {
        ch = fgetc(fptr);

        // Ignore newlines
        if (ch == *"\n") {
            continue;
        }
        
        // Toggle comment_ignore if there is a comment symbol
        if (ch == *"`") {
            comment_ignore = !comment_ignore;
            continue;
        }
        // Ignore following chars if comment_ignore is true
        if (comment_ignore) {
            continue;
        }

        // Enable/disable ignore_delimiters if there are parentheses
        if (ch == *"(") {
            ignore_delimiters = true;
            continue;
        } if (ch == *")") {
            ignore_delimiters = false;
            continue;
        }

        if (ch == *" ") {
            if (line_started) {
                continue;
            }
        } else {
            line_started = false;
        }


        if (!ignore_delimiters) {
            // Colons separate a type declaration from
            // the actual value. The syntax of a given
            // value to a command should be written as
            // the type followed by a colon and then the
            // value (e.g. int:4). If there is no type
            // declaration, the value is interpreted
            // as a string by default.

            if (ch == *":") {
                // Since the "chunk" variable is built up
                // character by character, if a colon is found
                // the chunk variable must contain the type name.

                chunk_encoding = 1;

                //printf("%s\n", segment_buffer);

                //chunk_type = match chunk.as_str() {
                //    "str" => 0, // String (0) (declaration optional)
                //    "int" => 1, // Integer (1)
                //    "float" => 2, // Float (2)
                //    "bool" => 3, // Boolean (3)
                //    _ => 4, // Default to string (0)
                //};

                // Since the chunk is built char by char, Clearing the chunk
                // at this point will make it so that when the delimiter code
                // below is triggered, the chunk will not contain the declaration
                // but the raw value only.
                segment_mem = 8;
                segment_len = 0;
                segment_buffer = malloc(segment_mem);

                continue;
            }


            // Delimiters are spaces and semicolons
            // A semicolon represents the end of a line
            // A space represents the end of a chunk unless between parentheses
            // This checks if the char is one of these delimiters: 
            if ((ch == *" ") || (ch == *";")) {
                // If a delimiter has been found, it means
                // the current chunk of the line should be processed
                // and added to the line buffer.

                // Allocate space in line for another segment pointer
                lineBuffer = realloc(lineBuffer, (segment_index+2)*sizeof(char*));
                lineBuffer[segment_index] = segment_buffer; // Add segment

                printf("%s %i ", segment_buffer, segment_index);

                // THIS IS WHERE CHUNKS SHOULD BE ENCODED BASED ON ENCODING.

                // Let's compile to C
                // When incl is seen, it should read a bzec (command file)
                // And add it as a function in the compiled C code





                // Set memory for next segment
                segment_mem = 8;
                segment_len = 0;
                segment_buffer = malloc(segment_mem);
                segment_index++;

                chunk_encoding = 0;


                // Append line if the delimiter is a semicolon
                if (ch == *";") {
                    // Reallocate to hold another line pointer 
                    lines = realloc(lines, (lineCount+1)*sizeof(char*));

                    // Add line sentinel
                    lineBuffer[segment_index] = lineSentinel;

                    // Append line to lines
                    lines[lineCount] = lineBuffer;

                    //if strcmp(lineBuffer[0], )

                    fprintf(out, "%s(", lineBuffer[0]);
                    printf("\n");

                    segment_index = 0;

                    // Setting the buffer to NULL leaves behind
                    // the memory address of the previous line
                    // so when realloc is called again, new
                    // memory will be created for the next line
                    // instead of overwriting the previous line
                    lineBuffer = NULL;
                    lineCount++;

                    line_started = true;
                }

                continue;
            }
        }


        // Append char to segment buffer

        //int len = stringLength(segment_buffer);

        // Allocate more memory for line segment if needed
        if (segment_mem < segment_len+1) {
            segment_mem += 8;
            segment_buffer = realloc(
                segment_buffer, segment_mem
            );
        }

        segment_buffer[segment_len] = ch;
        segment_buffer[segment_len + 1] = '\0';
        segment_len++;
    }


    // Add script sentinel
    char** scsentinel = malloc(1);
    scsentinel[0] = "SCPT_END";
    lines[lineCount] = scsentinel;

    fclose(fptr);
    fclose(out);

    return lines;
}